File: scala/src/main/scala/tds/td3/impl/PromotionWithDelegateImpl.scala
package tds.td3.impl

import tds.td3.api.{PromotionWithDelegate, Student}

// NON FUNCTIONAL: MUTATION
class PromotionWithDelegateImpl(id: Int, students: Set[Student]) extends PromotionImpl(id=id, students=students) with PromotionWithDelegate:

  private var _delegate: Option[Student] = None

  override def delegate: Option[Student] = _delegate

  // NON FUNCTIONAL: MUTATION
  override def setDelegate(student: Student): Unit =
    if (students.contains(student)) then _delegate = Option.apply(student)

  override def toString(): String =
    super.toString() + s" -- delegate: " + delegate.map(_.toString).getOrElse("none")

File: java/app/src/main/java/org/example/tds/td3/promowarn/version3/PromotionWithDelegateImpl.java
package org.example.td3.promowarn.version3;

import java.util.Optional;

public class PromotionWithDelegateImpl extends PromotionImpl implements PromotionWithDelegate {
    private Student delegate = null;

    public PromotionWithDelegateImpl(final int id) {
        super(id);
    }

    @Override
    public Optional<Student> delegate() {
        return Optional.ofNullable(delegate);
    }

    @Override
    public void chooseDelegate(final Student e) {
        if (students().contains(e)) {
            delegate = e;
        }
    }
}
